### PR TITLE
Cipher: add support for AES256 GCM tags

### PR DESCRIPTION
**Pull request by bmhatfield**
Related to the discussion on https://github.com/vapor/crypto/issues/68, this adds support for retrieving tags from an AES256GCM cipher, and passing them in to decrypt.

Note: Because GCM is technically a different interface, and because of the discussion on #68, I duplicated the relevant parts of Cipher and CipherAlgorithm into AuthenticatedCipher and AuthenticatedCipherAlgorithm.

While this results in some duplicated code (as noted on #68), the resulting interface is indeed much nicer - especially since you'll get a compile-time error if you mix and match unauthenticated and authenticated ciphers.

To help reduce some of the code duplication, we also created a pair of protocols - OpenSSLCipherAlgorithm and OpenSSLStreamCipher - that capture some of the needed common behavior, and deduplicates the code via protocol extensions.

Fixes https://github.com/vapor/crypto/issues/68

### CODE REVIEW

File: Sources/Crypto/Cipher/AuthenticatedCipher.swift [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+public final class AuthenticatedCipher: OpenSSLStreamCipher {
+    /// The `AuthenticatedCipherAlgorithm` (e.g., AES-256-GCM) to use.
+    public let algorithm: OpenSSLCipherAlgorithm
+
+    /// Internal OpenSSL `EVP_CIPHER_CTX` context.
+    public let ctx: UnsafeMutablePointer<EVP_CIPHER_CTX>
+
+    /// Byte length of a GCM tag
+    public static let gcmTagLength: Int = 16
```
Comment (MrMage) [inline Sources/Crypto/Cipher/AuthenticatedCipher.swift]: This is specific to AES-GCM, while the rest of the class isn't. Would there be a way around that?

File: Sources/Crypto/Cipher/AuthenticatedCipher.swift [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    ///
+    /// - parameters:
+    ///     - data: Plaintext data to encrypt.
+    ///     - key: Cipher key to use for encryption.
+    ///            This key must be an appropriate length for the cipher you are using. See `CipherAlgorithm.keySize`.
+    ///     - iv: Initialization vector to use for encryption.
+    ///           The IV must be an appropriate length for the cipher you are using. See `CipherAlgorithm.ivSize`.
+    /// - returns: Encrypted ciphertext and GCM tag.
+    /// - throws: `CryptoError` if reset, update, finalization or tag retrieval steps fail or if data conversion fails.
+    public func encrypt(_ data: LosslessDataConvertible, key: LosslessDataConvertible, iv: LosslessDataConvertible) throws -> (Data, Data) {
```
Comment (MrMage) [inline Sources/Crypto/Cipher/AuthenticatedCipher.swift]: nit: How about result type `(ciphertext: Data, tag: Data)`?

File: Sources/Crypto/Cipher/AuthenticatedCipher.swift [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+        return buffer
+    }
+
+    /// Sets the GCM Tag in the CIPHER_CTX struct.
+    ///
+    /// Note: This _must_ be called before `finish()` to set the tag.
+    ///
+    /// - throws: `CryptoError` if tag set fails
+    public func gcmTag(_ tag: Data) throws {
```
Comment (MrMage) [inline Sources/Crypto/Cipher/AuthenticatedCipher.swift]: nit: `{set,provide}GCMTag`? Also `{get,retrieve}GCMTag` above?

File: Sources/Crypto/Cipher/AuthenticatedCipher.swift [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            throw CryptoError.openssl(identifier: "EVP_CIPHER_CTX_ctrl", reason: "Failed setting tag (EVP_CTRL_GCM_SET_TAG).")
+        }
+    }
+
+    /// Cleans up and frees the allocated OpenSSL cipher context.
+    deinit {
+        EVP_CIPHER_CTX_cleanup(ctx)
+        EVP_CIPHER_CTX_free(ctx)
+    }
+
```
Comment (MrMage) [inline Sources/Crypto/Cipher/AuthenticatedCipher.swift]: super-nit: Remove this newline?

File: Tests/CryptoTests/CipherTests.swift [hunk 5]
```diff
@@ -42,6 +42,26 @@ class CipherTests: XCTestCase {
...
+    }
+
+    func testAES256GCMAuthenticationFailure() throws {
+        let message = "vapor"
+        let key = "passwordpasswordpasswordpassword"
+        let iv = "123456789012"
+        let (ciphertext, tag) = try AES256GCM.encrypt(message, key: key, iv: iv)
+        XCTAssertEqual(ciphertext.hexEncodedString(), "4fa166802c")
+
+        let badTag = tag.base64EncodedData()
```
Comment (MrMage) [inline Tests/CryptoTests/CipherTests.swift]: Suggestion: use `tag`, but change a single byte. With the current test, decryption might already fail due to an invalid tag length instead of a tag with the correct length, but an invalid value. Might also add a test where `tag` is shortened by a byte (see my comment in the implementation of `setTag`).

File: Sources/Crypto/Cipher/AuthenticatedCipher.swift [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    /// Sets the GCM Tag in the CIPHER_CTX struct.
+    ///
+    /// Note: This _must_ be called before `finish()` to set the tag.
+    ///
+    /// - throws: `CryptoError` if tag set fails
+    public func gcmTag(_ tag: Data) throws {
+        var buffer = tag
+
+        guard buffer.withMutableByteBuffer({ EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, Int32(AuthenticatedCipher.gcmTagLength), $0.baseAddress!) }) == 1 else {
```
Comment (MrMage) [inline Sources/Crypto/Cipher/AuthenticatedCipher.swift]: I think this will have unexpected results for a tag argument _shorter_ than `AuthenticatedCipher.gcmTagLength`. I would suggest checking for the exactly correct tag length here, and maybe adding a test that verifies that using a too short tag does not crash.

File: Tests/CryptoTests/CipherTests.swift [hunk 7]
```diff
@@ -42,6 +42,26 @@ class CipherTests: XCTestCase {
         try XCTAssertEqual(AES256.decrypt(ciphertext, key: key).convert(), message)
     }
 
+    func testAES256GCM() throws {
```
Comment (MrMage) [inline Tests/CryptoTests/CipherTests.swift]: Thanks for adding tests!

**Comment by bmhatfield** [comment]
@MrMage thanks for the feedback! I believe this PR is ready for another review!

**Comment by bmhatfield** [comment]
Oof, I take it back - looks like this doesn't compile on Linux due to how `EVP_AEAD_MAX_TAG_LENGTH` is defined.

**Comment by bmhatfield** [comment]
Okay, updated!

File: Sources/Crypto/Cipher/AuthenticatedCipher.swift [hunk 8]
```diff
@@ -117,11 +122,11 @@ public final class AuthenticatedCipher: OpenSSLStreamCipher {
 
         /// Require that the tag length is full-sized. Although it is possible to use the leftmost bytes of a tag,
         /// shorter tags pose both a buffer size risk as well as an attack risk.
-        guard buffer.count == EVP_AEAD_MAX_TAG_LENGTH else {
-            throw CryptoError.openssl(identifier: "EVP_CIPHER_CTX_ctrl", reason: "Tag length too short: \(buffer.count) != \(EVP_AEAD_MAX_TAG_LENGTH) (EVP_AEAD_MAX_TAG_LENGTH).")
+        guard buffer.count == AEAD_MAX_TAG_LENGTH else {
+            throw CryptoError.openssl(identifier: "EVP_CIPHER_CTX_ctrl", reason: "Tag length too short: \(buffer.count) != \(AEAD_MAX_TAG_LENGTH) (AEAD_MAX_TAG_LENGTH).")
```
Comment (MrMage) [inline Sources/Crypto/Cipher/AuthenticatedCipher.swift]: "Too short" is misleading, it might also be too long.

File: Tests/CryptoTests/CipherTests.swift [hunk 9]
```diff
@@ -55,11 +55,28 @@ class CipherTests: XCTestCase {
         let message = "vapor"
         let key = "passwordpasswordpasswordpassword"
         let iv = "123456789012"
-        let (ciphertext, tag) = try AES256GCM.encrypt(message, key: key, iv: iv)
+        var (ciphertext, tag) = try AES256GCM.encrypt(message, key: key, iv: iv)
+        XCTAssertEqual(ciphertext.hexEncodedString(), "4fa166802c")
+
+        // Forcibly overwrite bytes in the tag data to fail authentication
+        tag[2] = 0
```
Comment (MrMage) [inline Tests/CryptoTests/CipherTests.swift]: Micro-nit: I'd suggest setting just one byte to zero; that should be sufficient to make the test fail.

File: circle.yml [hunk 10]
```diff
@@ -117,7 +117,7 @@ workflows:
       - linux-vapor
       - linux-jwt
       - linux-mysql
-      - linux-postgresql
+      # - linux-postgresql
```
Comment (MrMage) [inline circle.yml]: Not sure what this does; seems unrelated to this commit?

Comment (bmhatfield) [reply circle.yml]: Commenting out that CI job disables the broken test job - there seems to be a CI-level misconfiguration there, so the "red X" Github shows about this PR is inaccurate. I can leave it on  if y'all prefer.

**Comment by bmhatfield** [comment]
@tanner0101 - I believe folks are deferring to you for final merge on this PR. If you have any questions that you'd like to chat out in real time, I hang out in the Vapor discord!

File: circle.yml (hunk 10 above)
Comment (bmhatfield) [reply circle.yml]: Removed that commit!

**Comment by bmhatfield** [comment]
@tanner0101 Ping on this PR - would love to see a merge!

File: Sources/Crypto/Cipher/CipherAlgorithm.swift [hunk 11]
```diff
@@ -40,37 +40,11 @@ public final class CipherAlgorithm {
     /// Only use this if you know what you are doing; use AES-256 GCM otherwise (see https://github.com/vapor/crypto/issues/59).
     public static let aes256cfb: CipherAlgorithm = .init(c: EVP_aes_256_cfb128())
 
-    /// AES-256 GCM cipher. This is the recommended cipher.
-    /// See the global `AES256GCM` constant on usage.
-    public static let aes256gcm: CipherAlgorithm = .init(c: EVP_aes_256_gcm())
```
Comment (tanner0101) [inline Sources/Crypto/Cipher/CipherAlgorithm.swift]: We need to offer deprecated support for this.

File: Sources/Crypto/Cipher/OpenSSLStreamCipher.swift [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+            return try data.withByteBuffer({ try closure($0) })
+        case .none:
+            return try closure(nil)
+        }
+    }
+}
+
+/// OpenSSLStreamCipher is a protocol representing a higher-level interface for managing various OpenSSL stream ciphers.
+public protocol OpenSSLStreamCipher {
+    func reset(key: LosslessDataConvertible, iv: LosslessDataConvertible?, mode: CipherMode) throws
```
Comment (tanner0101) [inline Sources/Crypto/Cipher/OpenSSLStreamCipher.swift]: All protocol methods must be documented (this counts toward the overall % documented for the framework)

File: Sources/Crypto/Cipher/CipherAlgorithm.swift (hunk 11 above)
Comment (MrMage) [reply Sources/Crypto/Cipher/CipherAlgorithm.swift]: No, we don't. This code has never actually worked in the first place. So it should never have been referenced anywhere.

Comment (vzsg) [reply Sources/Crypto/Cipher/CipherAlgorithm.swift]: Do we. It's the same as in the old days, GCM doesn't work without managing the tag. This constant was never tested (and would have never worked, as GCM forces you to do proper AEAD).

Comment (tanner0101) [reply Sources/Crypto/Cipher/CipherAlgorithm.swift]: Even just something like this could be better than removing it outright:

```swift
extension CipherAlgorithm {
    @available(*, unavailable, message: "Use AES256GCM type instead.")
    public static var aes256gcm: CipherAlgorithm { fatalError() }
}
```

Comment (bmhatfield) [reply Sources/Crypto/Cipher/CipherAlgorithm.swift]: It's worth noting that this module exposes a number of top-level "globals" like `AES256GCM`. Although the file that defines that constant has changed, the edge user experience has not (assuming you were using the globals). As far as I could tell, those "globals" were the way you access these algorithms.

So `public var AES256GCM: Cipher { return .init(algorithm: .aes256gcm) }` is now in `AuthenticatedCipher` instead. If you peep the tests, you can see that we can just call `AES256GCM.encrypt([...])` as if nothing had changed (except that it now works).

I think a compile time error for anyone who has manually instantiated that class would be better, in line with the thinking MrMage and vzsg have shared.

File: Sources/Crypto/Cipher/OpenSSLStreamCipher.swift (hunk 12 above)
Comment (bmhatfield) [reply Sources/Crypto/Cipher/OpenSSLStreamCipher.swift]: Sounds good - will do shortly.

File: Sources/Crypto/Cipher/CipherAlgorithm.swift (hunk 11 above)
Comment (tanner0101) [reply Sources/Crypto/Cipher/CipherAlgorithm.swift]: Using `@available(*, unavailable, message:)` does create a compile-time error. Just a tiny bit more descriptive. 

screen shot 2018-08-09 at 10 36 40 am

But I tend to agree that it's very unlikely anyone was using that directly.

**Comment by vzsg** [comment]
TIL that the doc comments in the Cipher class push people to use GCM, and now we have the first user who tried to do so, only to fail.

**Comment by MrMage** [comment]
Sorry, I think I was the one writing that comment. However, AES-GCM is a good thing (when it works), so once this PR is merged the comments should actually make sense.

File: Sources/Crypto/Cipher/AuthenticatedCipher.swift [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+///     let (ciphertext, tag) = try AES256GCM.encrypt("vapor", key: key, iv: iv)
+///     print(ciphertext) // Encrypted Data
+///     AES256GCM.decrypt(ciphertext, key: key, iv: iv, tag: tag).convert(to: String.self) // "vapor"
+///
+public var AES256GCM: AuthenticatedCipher { return .init(algorithm: .aes256gcm) }
+
+/// Max Tag Length. Used for defining the size of input and output tags.
+///     Redefined from OpenSSL's EVP_AEAD_MAX_TAG_LENGTH, which seems to be improperly defined on some platforms.
+///     You can find the original #define here: https://github.com/libressl/libressl/blob/master/src/crypto/evp/evp.h#L1237-L1239
+public let AEAD_MAX_TAG_LENGTH: Int32 = 16
```
Comment (tanner0101) [inline Sources/Crypto/Cipher/AuthenticatedCipher.swift]: Can we keep this internal?

File: Sources/Crypto/Cipher/AuthenticatedCipher.swift [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+
+public final class AuthenticatedCipher: OpenSSLStreamCipher {
```
Comment (tanner0101) [inline Sources/Crypto/Cipher/AuthenticatedCipher.swift]: Need a comment on this public type for 100% docs coverage.

File: Sources/Crypto/Cipher/AuthenticatedCipher.swift [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+        try update(data: data, into: &buffer)
+        try setTag(tag)
+        try finish(into: &buffer)
+
+        return buffer
+    }
+
+    /// Gets the Tag from the CIPHER_CTX struct.
+    ///
+    /// Note: This _must_ be called after `finish()` to retrieve the generated tag.
```
Comment (tanner0101) [inline Sources/Crypto/Cipher/AuthenticatedCipher.swift]: Using `- note: ` will cause this to render properly in the Xcode popup.

File: Sources/Crypto/Cipher/AuthenticatedCipher.swift [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+        guard buffer.withMutableByteBuffer({ EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, AEAD_MAX_TAG_LENGTH, $0.baseAddress!) }) == 1 else {
+            throw CryptoError.openssl(identifier: "EVP_CIPHER_CTX_ctrl", reason: "Failed getting tag (EVP_CTRL_CCM_GET_TAG).")
+        }
+
+        return buffer
+    }
+
+    /// Sets the Tag in the CIPHER_CTX struct.
+    ///
+    /// Note: This _must_ be called before `finish()` to set the tag.
```
Comment (tanner0101) [inline Sources/Crypto/Cipher/AuthenticatedCipher.swift]: Same here.

File: Sources/Crypto/Cipher/OpenSSLCipherAlgorithm.swift [hunk 17]
```diff
@@ -0,0 +1,37 @@
+import CNIOOpenSSL
+
+/// OpenSSLCipherAlgorithm represents a common set of properties shared by
+/// OpenSSL cipher algorithms.
+public protocol OpenSSLCipherAlgorithm {
+    var c: UnsafePointer<EVP_CIPHER> { get }
+    init(c: UnsafePointer<EVP_CIPHER>)
+
+    var type: Int32 { get }
```
Comment (tanner0101) [inline Sources/Crypto/Cipher/OpenSSLCipherAlgorithm.swift]: All protocol members require doc block for 100% docs coverage. I recommend just copying the doc blocks you have from the implementation, then replacing the doc blocks there with:

```swift
/// See `OpenSSLCipherAlgorithm`
```

That way we prevent doc duplication and always point people to the most accurate docs.

File: Sources/Crypto/Cipher/OpenSSLStreamCipher.swift [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+    /// Encrypts arbitrary data to encrypted ciphertext.
+    case encrypt = 1
+
+    /// Decrypts encrypted ciphertext back to its original value.
+    case decrypt = 0
+}
+
+/// Wrapper to allow for safely working with a potentially-nil Data's byte buffer.
+extension Optional where Wrapped == Data {
+    // Note: It's iffy to try this with a mutable buffer, so an Optional version
```
Comment (tanner0101) [inline Sources/Crypto/Cipher/OpenSSLStreamCipher.swift]: `- note: ` syntax here, too.

**Comment by bmhatfield** [comment]
@tanner0101 I believe I have addressed all of your PR feedback, ready for a re-review.

**Comment by penny-coin** [comment]
Hey @bmhatfield, you just merged a pull request, have a coin! 

You now have 1 coins.
